core/src/main/java/org/apache/calcite/runtime/FlatLists.java

```diff
@@ -574,7 +574,7 @@ public boolean equals(Object o) {
             && Objects.equals(this.t1, that.t1)
             && Objects.equals(this.t2, that.t2);
       }
-      return o.equals(this);
+      return Arrays.asList(t0, t1, t2).equals(o);
```

Comment:
Maybe better define a static variable to store `Arrays.asList(t0, t1, t2)`

Reply:
On the other hand, can we just return `false` here?

Reply:
Thanks. I'll abandon this version, and hope for someone else to finish it while I'm on vacation:)